### PR TITLE
Validate non-deterministic generated columns if provided on primary

### PR DESCRIPTION
**Pull request by mfussenegger**
Fixes https://github.com/crate/crate/issues/14218

### CODE REVIEW

**Comment by matriv** [comment]
I'm leaning towards the fix, and have deterministic and non-deterministic gen cols behave consistently regarding validation.

Could you please explain a bit more this: ?
> Currently it's not really possible to easily copy tables with something like `created ... current_timestamp`, if we kept things as they are it would work.

If the source table has a gen col with `current_timestamp` then the column is filled in, so it can be copied to the target new table. You'd probably mean the case that you don't have the col on source table and you have a generated column on the target one?

**Comment by mfussenegger** [comment]
> If the source table has a gen col with current_timestamp then the column is filled in, so it can be copied to the target new table. You'd probably mean the case that you don't have the col on source table and you have a generated column on the target one?

If you have:

```sql
create table tsrc (x int, ts timestamp with time zone as CURRENT_TIMESTAMP)
create table tdst (x int, ts timestamp with time zone as CURRENT_TIMESTAMP)
insert into tdst (x, ts) (select x, ts from tsrc)
```

It will never be able to insert any records.
(If the source has a generated column or not doesn't matter)

Currently a common use case could be to copy the table to modify some columns, either name, type, or dropping it.

As @hlcianfagna pointed out one can use a `default` clause instead if the use case is to have a `created` column.
But if the use case is a `last_update` column it gets fuzzier. If we _didn't_ do the validation, we would leave it up to the user to control the behavior:

```
create table tdst (x int, last_update timestamp with time zone as CURRENT_TIMESTAMP)

-- preserve last_update time from original table
insert into tdst (x, last_update) (select x, ts from tsrc)

-- Re-compute last_update time
insert into tdst (x) (select x from tsrc)
```

With validation there is no way to preserve the original times.


tl;dr, as far as I can tell:

Advantages of not-validating:

- Users can control behavior and copy tables

Disadvantages

- Consistency. Although even this is debatable, because one could redefine the validation as something like "ensures the provided value fits in the set of values the random expression can return" and than it is consistent for both cases. (E.g. we could mark if a scalar can return null, and then prevent null insertion if scalars never return null)


So to me the question is if there are any significant downside in _not_ having the validation.

**Comment by hlcianfagna** [comment]
For a `last_update` column, the user may want to import the `last_update` timestamps from the source table or get fresh `last_update` timestamps so a default would work as it allows that flexibility. I also usually see `last_update` columns implemented either client-side or with triggers, not as generated columns.

**Comment by matriv** [comment]
I think the behaviour described is a valid use case, users may want to copy over existing timestamps from old table and from then on get the generated timestamp for new inserts.
On the other hand I don't like to completely drop validation and end up with a table that is not consistent (e.g. missing values for a generated col).
How about making this configurable (relax the validation based on a session param, by default is on)?

**Comment by seut** [comment]
I think we should reconsider our differentiation of generated columns vs. default columns to clarify expectations and also to align with other DB's.
Such as, providing a concrete value for a generated column would be forbidden at all (PostgreSQL even do not allow the usage of non-deterministic functions) (Current PG and CockroachDB behaviour).
Of course this would be a breaking change and we'd need to carefully document this and point the user to use a `DEFAULT` clause instead for such use-cases.
As @hlcianfagna pointed out, the `last_updated` use case would then work fine with a `DEFAULT` expression.

For now, we may get the old behaviour back and follow-up with the breaking change later on?

**Comment by mfussenegger** [comment]
A default clause doesn't work for a `last_update` use case because you want the value to re-compute on each update.  `default` only applies on the initial insert if the column is absent.

**Comment by hlcianfagna** [comment]
I just notice the behaviour on https://github.com/crate/crate/issues/7628 persists on the latest nightly, I find it quite counter-intuitive for a column that is stored and indexed and not a virtual column computed on demand.
The `ON UPDATE` MySQL feature on the other side looks interesting, it could be generalised to say `ON UPDATE` of the record this field is updated recomputing the given expression and that would cover many of the use cases of triggers.

**Comment by matriv** [comment]
I agree that inconvenience for such cases is more important than consistency, so fix it for 5.3 and then introduce a breaking change.

How about making this configurable, though?

**Comment by mfussenegger** [comment]
> So I'd vote for changing this as a breaking change while merging this as a fix for now.

We could also change the release notes for 5.3 to mention this change as the damage is already kinda done.
If we apply this change, and break it for 5.4, we have 3 breaking changes, from 5.2 to 5.3.0, then from 5.3.x to 5.3.x+1 where we include this PR, and again with 5.3 to 5.4.0.

Is there an advantage to doing it in 5.4.0 that I'm missing?

> How about making this configurable, though?

I think we could consider this if users request it. If nobody is asking for it, it's unnecessary complexity.

**Comment by proddata** [comment]
There is another issue connected to this one on partitioned tables:
https://github.com/crate/crate/issues/14304

**Comment by mfussenegger** [comment]
https://github.com/crate/crate/pull/14323
